torch/_refs/nn/functional/__init__.py

```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
+    """
+
+    if weight is None:
+        current_weight = torch.where(
+            ignore_classes_mask,
+            ignore_class_weight,
+            default_class_weight,
+        )
+    else:
+        current_weight = torch.where(
+            ignore_classes_mask,
+            ignore_class_weight.expand_as(flat_target),
+            weight[flat_target],
+        )
+
+    if input.ndim == 1:
+        loss = -input[target] * current_weight
+    elif input.ndim == 2:
+        batch_size = input.shape[0]
+        loss = -input[torch.arange(batch_size), target] * current_weight
+    else:
+        batch_size = input.shape[0]
+        height = input.shape[2]
+        width = input.shape[3]
+        extent = height * width
+        numel = batch_size * extent
+        bdx = torch.arange(numel) // extent
+        hdx = (torch.arange(numel) - (bdx * extent)) // width
+        wdx = torch.arange(numel) % width
+        loss = -input[bdx, flat_target, hdx, wdx] * current_weight
+    loss = torch.reshape(loss, target.shape)
+
+    if reduction == "none":
+        return loss
+    elif reduction == "sum":
+        return torch.sum(loss)
+    else:
+        return torch.sum(loss) / torch.sum(current_weight)
+
+
+@register_decomposition(torch.ops.aten.nll_loss)
+def nll_loss(
+    input: TensorLikeType,
+    target: TensorLikeType,
+    weight: Optional[TensorLikeType] = None,
+    size_average: Optional[bool] = None,
+    ignore_index: int = -100,
+    reduce: Optional[bool] = None,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    if size_average is not None or reduce is not None:
+        # TODO: raise exception instead of converting value
+        # msg = "size_average and reduce args are deprecated, please use reduction argument."
+        reduction = _get_string_reduction_arg(size_average=size_average, reduce=reduce)
+
+    if input.ndim == 3 or input.ndim > 4:
+        # input ndim is == 3 or > 4
+        batch_size = input.shape[0]
+        num_classes = input.shape[1]
+        out_size = [batch_size] + list(input.shape[2:])
+
+        if target.shape[1:] != input.shape[2:]:
+            msg = "Expected target size {} but got {}"
+            raise ValueError(msg.format(out_size, target.shape))
+
+        # support empty batches, see #15870
+        if input.numel() > 0:
+            input = torch.reshape(input, [batch_size, num_classes, 1, -1])
+        else:
+            input = torch.reshape(input, [batch_size, num_classes, 0, 0])
+
+        if target.numel() > 0:
+            target = torch.reshape(target, [batch_size, 1, -1])
+        else:
+            target = torch.reshape(target, [batch_size, 0, 0])
+
+        if reduction == "none":
+            return _nll_loss_nd(input, target, weight, reduction, ignore_index)
+        else:
+            result = _nll_loss_nd(input, target, weight, reduction, ignore_index)
+            return torch.reshape(result, out_size)
+    else:
```

Comment:
Add a comment describing what this else branch is for.

From a code organization and readability standpoing these branches seem a little odd. Maybe we can explain them better?

In particular -- can input be zero or one dimension? If so, how do we interpret that? The documentation for suggests that input should have at least two dimensions. And why are inputs with three or four dimensions special? 

Finally, prefer putting shorter branches which short-circuit first. That typically lets code have fewer indentation levels:

```
# shortcircuits if foo because...
if foo:
  return x

# implicit else branch doesn't have to be indented
...
```

Reply:
I refactored _nll_loss_nd to handle 1-3 dimensions. If there are more than 3 dimensions, the k-dimension is flattened to create a 3D tensor. The Aten implementation used a 4D case for image inputs.
```
    # The _nll_loss_nd helper function handles the most common cases.
    # ndim == 1 (Single Example)
    #   => Batch Size: 1, Input: (C), Target: ()
    # ndim == 2 (k = 1)
    #   => Batch Size: N, Input: (N, C), Target: (N)
    # ndim == 3 (k > 1)
    #   => Batch Size: N, Input: (N, C, K), Target: (N, K)
    # ndim > 3
    #   => reshape the input and target to the 3-D case
```

Reply:
Is the 4D case interesting to model here?